jvm-packages/xgboost4j-spark/src/main/scala/org/apache/spark/SparkParallelismTracker.scala

```diff
@@ -0,0 +1,122 @@
+ limitations under the License.
+ */
+
+package org.apache.spark
+
+import java.net.URL
+
+import org.apache.commons.logging.LogFactory
+import org.apache.spark.scheduler.{SparkListener, SparkListenerExecutorRemoved}
+import org.codehaus.jackson.map.ObjectMapper
+
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future, TimeoutException}
+
+/**
+ * A tracker that ensures enough number of executor cores are alive.
+ * Throws an exception when the number of alive cores is less than nWorkers.
+ *
+ * @param sc The SparkContext object
+ * @param timeout The maximum time to wait for enough number of workers.
+ * @param nWorkers nWorkers used in an XGBoost Job
+ */
+class SparkParallelismTracker(
+    val sc: SparkContext,
+    timeout: Long,
+    nWorkers: Int) {
+
+  private[this] val mapper = new ObjectMapper()
+  private[this] val logger = LogFactory.getLog("XGBoostSpark")
+  private[this] val url = sc.uiWebUrl match {
+    case Some(baseUrl) => new URL(s"$baseUrl/api/v1/applications/${sc.applicationId}/executors")
+    case _ => null
+  }
+  private[spark] var failedReason: String = null
+
+  private[this] def numAliveCores: Int = {
+    try {
+      mapper.readTree(url).findValues("totalCores").asScala.map(_.asInt).sum
+    } catch {
+      case ex: Throwable =>
+        logger.warn(s"Unable to read total number of alive cores from REST API." +
+          s"Health Check will be ignored.")
+        ex.printStackTrace()
+        Int.MaxValue
+    }
+  }
+
+  private[this] def waitForCondition(
+      condition: => Boolean,
+      timeout: Long,
+      checkInterval: Long = 100L) = {
+    val monitor = Future {
+      while (!condition) {
+        Thread.sleep(checkInterval)
+      }
+    }
+    Await.ready(monitor, timeout.millis)
+  }
+
+  private[this] def safeExecute[T](body: => T): T = {
+    sc.listenerBus.listeners.add(0, new ExecutorLostListener(this))
+    try {
+      val result = body
+      if (failedReason != null) {
+        throw new SparkException(s"ExecutorLost during XGBoost Training: $failedReason")
+      }
+      result
+    } finally {
+      sc.listenerBus.listeners.remove(0)
+      failedReason = null
+    }
+  }
+
+  /**
+   * Execute a blocking function call with two checks on enough nWorkers:
+   *  - Before the function starts, wait until there are enough executor cores.
+   *  - During the execution, throws an exception if there is any executor lost.
+   *
+   * @param body A blocking function call
+   * @tparam T Return type
+   * @return The return of body
+   */
+  def execute[T](body: => T): T = {
+    if (timeout <= 0) {
+      body
+    } else {
+      try {
+        waitForCondition(numAliveCores >= nWorkers, timeout)
+      } catch {
+        case _: TimeoutException =>
+          throw new SparkException(s"Unable to get $nWorkers workers for XGBoost training")
```

Comment:
SparkException might not be a good choice as this is used to describe the behavior in Spark, you can use  `IllegalStateException`